BLE_GAP/source/main.cpp

```diff
@@ -0,0 +1,494 @@
+
+        _event_queue.call_in(2000, &_ble.gap(), &Gap::disconnect,
+                             Gap::LOCAL_HOST_TERMINATED_CONNECTION);
+    };
+
+    /** This is called by Gap to notify the application we disconnected,
+     *  in our case it calls demo_mode_end() to progress the demo */
+    void on_disconnect(const Gap::DisconnectionCallbackParams_t *event)
+    {
+        printf("Disconnected\r\n");
+
+        /* we have successfully disconnected ending the demo, move to next mode */
+        _event_queue.call(this, &GAPDevice::demo_mode_end);
+    };
+
+    /** called if timeout is reached during advertising or scanning */
+    void on_timeout(const Gap::TimeoutSource_t source)
+    {
+        _demo_duration.stop();
+
+        switch (source) {
+            case Gap::TIMEOUT_SRC_ADVERTISING:
+                printf("Stopped advertising early due to timeout parameter\r\n");
+                break;
+            case Gap::TIMEOUT_SRC_SCAN:
+                printf("Stopped scanning early due to timeout parameter\r\n");
+                break;
+            case Gap::TIMEOUT_SRC_CONN:
+                printf("Failed to connect after scanning %d advertisements\r\n", _scan_count);
+                _event_queue.call(this, &GAPDevice::print_performance);
+                _event_queue.call(this, &GAPDevice::demo_mode_end);
+                break;
+            default:
+                printf("Unexpected timeout\r\n");
+                break;
+        }
+    };
+
+    /** clean up after last run, cycle to the next mode and launch it */
+    void demo_mode_end()
+    {
+        /* reset the demo ready for the next mode */
+        _scan_count = 0;
+        _demo_duration.stop();
+        _demo_duration.reset();
+
+        /* cycle through all demo modes */
+        _set_index++;
+
+        /* switch between advertising and scanning when we go
+         * through all the params in the array */
+        if (_set_index >= (_is_scanning? SCAN_PARAM_SET_MAX : ADV_PARAM_SET_MAX)) {
+            _set_index = 0;
+            _is_scanning = !_is_scanning;
+        }
+
+        _ble.shutdown();
+        _event_queue.break_dispatch();
+    };
+
+    /** print some information about our radio activity */
+    void print_performance()
+    {
+        /* measure time from mode start, may have been stopped by timeout */
+        uint16_t duration = _demo_duration.read_ms();
+
+        if (_is_scanning) {
+            /* convert ms into timeslots for accurate calculation as internally
+             * all durations are in timeslots (0.625ms) */
+            uint16_t interval_ts = GapScanningParams::MSEC_TO_SCAN_DURATION_UNITS(
+                    scanning_params[_set_index].interval);
+            uint16_t window_ts = GapScanningParams::MSEC_TO_SCAN_DURATION_UNITS(
+                    scanning_params[_set_index].window);
+            uint16_t duration_ts = GapScanningParams::MSEC_TO_SCAN_DURATION_UNITS(
+                    duration);
```

Comment:
Indentation of 4. Parenthesis on next line, not indented; Or put duration on the same line as the function name.